Fix a race condition in supported groups and ciphers handling

mattcaswell opened the pull request:
In TLSv1.3 the supported groups can be negotiated each time a handshake
occurs, regardless of whether we are resuming or not. We should not store
the supported groups information in the session because session objects
can be shared between multiple threads and we can end up with race
conditions. For most users this won't be seen because, by default, we
use stateless tickets in TLSv1.3 which don't get shared. However if you
use SSL_OP_NO_TICKET (to get stateful tickets in TLSv1.3) then this can
happen.

The answer is to move the supported the supported group information into
the SSL object instead.

ayoubserti commented:
Hi,
Thank @mattcaswell 
I think it's not only about supported group. I also see the crash happen with ciphers selection 
`statem_srvr.c`

```
if (!s->hit || SSL_IS_TLS13(s)) {
        sk_SSL_CIPHER_free(s->session->ciphers);
        s->session->ciphers = ciphers;
        if (ciphers == NULL) {
            SSLfatal(s, SSL_AD_INTERNAL_ERROR,
                     SSL_F_TLS_EARLY_POST_PROCESS_CLIENT_HELLO,
                     ERR_R_INTERNAL_ERROR);
            goto err;
        }
        ciphers = NULL;
    }
```
It's way I suggest to lock session when it used by one thread.
I let you know when applying the patch.

mattcaswell commented:
> I think it's not only about supported groups. I also see the crash happen within ciphers selection

Any other places or just those two?

mattcaswell commented:
I added an additional commit to fix the other instance.

> It's way I suggest to lock session when it used by one thread.

There really should be no need for that since the session object is supposed to be immutable. If we find cases where it isn't, then that's a bug.

ayoubserti commented:
> > I think it's not only about supported groups. I also see the crash happen within ciphers selection
> 
> Any other places or just those two?
I'm not sure.
I keep you informed when applying your patch.

ayoubserti commented:
Hello,
no more crash using mattcaswell:supp-group-race branch.
Thank you

mattcaswell commented:
Great - thanks for testing it.

t8m commented on ssl/ssl_locl.h:
```diff
@@ -570,9 +569,7 @@ struct ssl_session_st {
         size_t ecpointformats_len;
         unsigned char *ecpointformats; /* peer's list */
```
Would it make sense to move also the ecpointformats?

mattcaswell replied:
AFAICT there is no race condition with ecpointformats so it doen't *need* to move. However, there isn't a good reason to store this data in the session object, and therefore for consistency reasons it probably does make sense to move this one too.

mattcaswell commented:
I rebased this to fix a merge conflict with master, and added an additional commit to also move the ecpointformats data out of the session.

Ping @t8m.

mattcaswell commented:
Pushed to master. Unfortunately there was a non-trivial conflict when I tried to backport this to 1.1.1, so I have opened a new PR for the backport. See #9176.

kaduk commented:
It seems like this presents a slight behavior change, in that (e.g.) SSL_get_client_ciphers() will return an empty list after pre-TLS 1.3 resumption handshakes?  It's not clear that this was a terribly useful thing to do, but I want to check that we have considered those effects and are willing to accept them.

kaduk commented on ssl/s3_lib.c:
```diff
@@ -3586,8 +3586,8 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
 
             if (!s->session)
                 return 0;
```
It seems pretty likely that this ```if (!s->session)``` was just to avoid a NULL dereference; it may well be stale logic at this point.  (That said, it's probably harmless, since we will definitely have an attached session by the point that we're parsing the peer's group list.)

richsalz replied:
if the line is kept, make it explicit test against NULL
